kubernetes/kubernetes.go

```diff
@@ -206,6 +206,22 @@ func (in *K8SClient) IsOpenShift() bool {
 	return *in.isOpenShift
 }
 
+func (in *K8SClient) IsGatewayAPI() bool {
+	if in.GatewayAPI() == nil {
+		return false
+	}
+	if in.isGatewayAPI == nil {
+		isGatewayAPI := false
+		// GatewayClass should exist got K8s Gateway API to work.
+		_, err := in.GatewayAPI().GatewayV1alpha2().GatewayClasses().List(in.ctx, emptyListOptions)
```

Comment:
Here any error reason means Gateway API is not supported. So probably better to log any error different from NotFound.

Reply:
Well it could be that Kiali is just having temporary problems talking to the kube API. We could retry for some subset of errors like `IsTimeout` or `IsTooManyRequests`. Another scenario is that someone could've deployed only _some_ of the CRDs but not all e.g. `GatewayClasses` were deployed but not `Gateways` or `HTTPRoutes`. I don't think we need to overcomplicate things though and probably just checking for one or two errors is good enough.

I actually get `Forbidden` errors when the Gateway API CRDs are not present which could either mean that the CRDs do not exist _or_ that the CRDs do exist and Kiali doesn't have access to them. Since we can't differentiate between the two scenarios, what do you think about just logging a `Warning` level log in case of any error?

Reply:
Have you tried this when Kiali doesn't have cluster scoped permissions? i.e. `accessible_namespaces != **`? The cluster scoped list may always fail in this scenario.

Reply:
Seeing a bunch of:
```
2022-09-20T14:38:47Z ERR Error checking Kubernetes Gateway API configuration: gatewayclasses.gateway.networking.k8s.io is forbidden: User "system:serviceaccount:istio-system:kiali-service-account" cannot list resource "gatewayclasses" in API group "gateway.networking.k8s.io" at the cluster scope
```
when you visit the apps list page. This looks like the only place where we list `GatewayClasses`.

Reply:
I think the reason I'm seeing multiple error messages is that Kiali creates multiple `K8sClient`. The layer creates one, the cache creates another, and the client factory creates them as needed. Having the `IsGatewayAPI` check on the `K8sClient` means that check will happen each time a new client is initialized which happens more than once in Kiali. That's probably fine and restarting kiali once you deploy the CRDs into the cluster will still fix everything.

The bigger issue is the cluster scoped list call to GatewayClasses. With `spec.deployment.accessible_namespaces = bookinfo`:
```
kubectl auth can-i list gatewayclasses --as system:serviceaccount:istio-system:kiali-service-account
> Warning: resource 'gatewayclasses' is not namespace scoped in group 'gateway.networking.k8s.io'
> no
kubectl auth can-i list gatewayclasses --as system:serviceaccount:istio-system:kiali-service-account -n bookinfo
> Warning: resource 'gatewayclasses' is not namespace scoped in group 'gateway.networking.k8s.io'
> yes
```

One potential workaround is to do a namespaced scoped call using the `Config.IstioNamespace`. This will work for operator deployments since the operator automatically gives Kiali a role for this namespace but I don't think that happens automatically for helm deployments. Another option is to use one of the namespaces in `spec.deployment.accessible_namespaces`. These can be regexs though so you might need to first use the `NamespaceService` to get a list of name of accessible namespaces. Unless there's another way test to see if the CRDs have been deployed?

Reply:
I had tested both cases accessible_namespaces="**" and accessible_namespaces = "bookinfo" on OCP installed kiali via operator and deployed on local machine. When CRD is installed or not. In all cases I did not see any error "Error checking Kubernetes Gateway API configuration....", even after listing apps. Not sure how it could happen on your env.
Only case when bunch of errors logged is the case when CRD is removed and before Kiali pod is restarted.
So at this point I will suggest to leave Gateway API recognition as it is,

Reply:
> So at this point I will suggest to leave Gateway API recognition as it is, only change the Error log level to Warn, not to make confusion, then wait for community feedback. As Gateway API itself is still in WIP.

My only concern is that for users who set `accessible_namespaces` to something other than `**`, Kiali will be logging error messages about gateway APIs even if they deploy the CRDs. With most of the info coming from the registry anyways, I don't think it'll impact the user experience much for right now with the gateway api info only being listed in the config.